opnode/l1/downloader_test.go

```diff
@@ -0,0 +1,206 @@
+package l1
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"errors"
+	"fmt"
+	"math/big"
+	"math/rand"
+	"testing"
+
+	"github.com/ethereum-optimism/optimistic-specs/opnode/eth"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/trie"
+	"github.com/stretchr/testify/assert"
+)
+
+type retryReceipt struct {
+	tries   int
+	receipt *types.Receipt
+}
+type mockDownloaderSource struct {
+	block    *types.Block
+	receipts map[common.Hash]*retryReceipt
+}
+
+func (m *mockDownloaderSource) BlockByHash(ctx context.Context, hash common.Hash) (*types.Block, error) {
+	if m.block == nil {
+		return nil, errors.New("no block here")
+	}
+	return m.block, nil
+}
+
+func (m *mockDownloaderSource) TransactionReceipt(ctx context.Context, txHash common.Hash) (*types.Receipt, error) {
+	ret, ok := m.receipts[txHash]
+	if !ok {
+		return nil, errors.New("no receipt here")
+	}
+	ret.tries -= 1
+	if ret.tries < 0 {
+		return ret.receipt, nil
+	}
+	return nil, fmt.Errorf("receipt takes %d more tries to download", ret.tries)
+}
+
+func RandomL1Block(txCount int) (*types.Block, []*types.Receipt) {
```

Comment:
This seems like a very useful function to have in some kind of test util library! Similarly with its subparts (e.g. we could have `RandomTx`, etc).

Reply:
Not too sure, we mostly just care about the receipts. Was even thinking of embedding empty/nil transactions to simplify it. Hiding the block data behind an interface, and only dealing with the minimal instead of generating a full block, would be nice.